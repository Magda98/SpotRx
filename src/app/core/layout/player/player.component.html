@if(player) {
<div class="player">
  <ng-container>
    @if(playerService.playbackState | async; as position){
    <mat-slider class="song-slider" min="1" [max]="player.duration" step="10">
      <input (dragEnd)="setPosition($event)" [value]="position" matSliderThumb>
    </mat-slider>
    }
    <div class="song-info">
      <img [alt]="player.track_window.current_track.album.name" class="album-image"
        [src]="player.track_window.current_track.album.images[0].url">
      <div class="album-artist">
        <h2 class="name">{{player.track_window.current_track.name}}</h2>
        <span class="album-name">
          {{player.track_window.current_track.album.name}}
        </span>
        <div class="artist-container">
          @for(item of player.track_window.current_track.artists; track item.name){
          <span class="artist">{{item.name}}</span>
          }
        </div>
      </div>
    </div>
    <div class="action-buttons-container">
      <div class="action-buttons">
        <button appButton type="button" aria-label="previous" [mode]="'text'" (click)="playerService.skipPrev()">
          <mat-icon>skip_previous</mat-icon>
        </button>
        <button appButton type="button" aria-label="play" [mode]="'text'" (click)="playerService.tooglePlay()">
          @if(player.paused){
          <mat-icon>play_circle_outline</mat-icon>
          }
          @else{
          <mat-icon>pause_circle_outline</mat-icon>
          }
        </button>
        <button appButton type="button" aria-label="next" [mode]="'text'" (click)="playerService.skipNext();">
          <mat-icon>skip_next</mat-icon>
        </button>
      </div>
    </div>
    <div class="slider-container">
      <button appButton type="button" [mode]="'text'">
        <app-icon [ngClass]="{'primary': player.shuffle }" (click)="toggleShuffle()" class="icon"
          type="shuffle"></app-icon>
      </button>
      <button appButton type="button" [mode]="'text'" (click)="toggleSavedTrack()">
        <app-icon [ngClass]="{'primary': isSavedTrack() === true }" class="icon" type="heart"></app-icon>
      </button>
      <mat-slider class="volume-slider" min="0" max="1" step="0.1">
        <input (dragEnd)="setVolume($event)" value="0.5" matSliderThumb>
      </mat-slider>
    </div>
  </ng-container>
</div>
}