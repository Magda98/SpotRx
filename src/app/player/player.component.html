<div class="player">
    <!-- <mat-slider min="1" max="5" step="0.5" value="1.5"></mat-slider> -->
    <button *ngIf="playerState | async as player" class="play" (click)="togglePlay()">
        <mat-icon *ngIf="player.paused">play_circle_outline</mat-icon>
        <mat-icon *ngIf="!player.paused">pause_circle_outline</mat-icon>
        {{player.position}}{{player.paused}}
    </button>

</div>