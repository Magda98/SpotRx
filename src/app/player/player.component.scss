.player {
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  width: 100%;
  height: 140px;
  background-color: #2c2f31;
  padding-left: 0;
  box-sizing: border-box;
  display: grid;
  grid-template-columns: 1fr;
  padding-top: 16px;

  @media screen and (min-width: 980px) {
    grid-template-columns: 1fr 1fr 1fr;
    padding-left: 220px;
  }

  .song-slider {
    position: absolute;
    top: -23px;
    left: 0;
    right: 0;

    @media screen and (min-width: 980px) {
      left: 212px;
    }
  }

  .action-buttons {
    display: flex;
    justify-content: space-between;
    align-items: center;
    margin-left: 0;
    padding: 0 8px;

    @media screen and (min-width: 980px) {
      margin-left: 25%;
      width: 100%;
      max-width: 250px;
    }
  }

  .song-info {
    display: flex;
    align-items: center;
    padding: 0 8px;

    .album-image {
      width: 60px;
      height: 60px;
      object-fit: cover;
      border-radius: 5px;
      display: block;
      margin-right: 10px;

      @media screen and (min-width: 980px) {
        width: 80px;
        height: 80px;
      }
    }

    .name {
      margin-bottom: 0;
    }

    .artist-container {
      display: flex;
      opacity: 0.6;
      margin-top: 5px;
    }

    .artist {
      margin-right: 5px;
    }

    .album-artist {
      display: flex;
      flex-direction: column;
    }

    .album-name {
      opacity: 0.6;
      font-size: 12px;

      @media screen and (min-width: 980px) {
        font-size: 16px;
      }
    }
  }

  .slider-container {
    display: flex;
    justify-content: center;
    align-items: center;
  }

  .volume-slider {
    display: none;

    @media screen and (min-width: 980px) {
      display: block;
      max-width: 150px;
      width: 100%;
    }
  }
}
