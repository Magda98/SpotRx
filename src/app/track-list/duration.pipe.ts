import { Pipe, PipeTransform } from '@angular/core';

@Pipe({
  name: 'duration'
})
export class DurationPipe implements PipeTransform {

  transform(duration: number): string {
    const temp = duration / (1000);
    const m = Math.floor(temp / 60);
    const s = Math.floor((temp % 60));
    
    const durationString = s < 10? `${m}:0${s}` : `${m}:${s}`
    return durationString;
  }

}
