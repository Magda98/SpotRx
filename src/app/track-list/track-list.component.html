<div class="track-list-container">
  <mat-paginator class="paginator" (page)="handlePageEvent($event)" [length]="total()" [pageSize]="pageSize"
    [pageIndex]="currentPage()" aria-label="Select page">
  </mat-paginator>
  @if(tracksListQuery().isSuccess()){
  <ul class="track-list">
    @for(item of tracksListQuery().data()?.items; track item.track.id; let i = $index){
    <li (click)="play(i)" class="track">
      <div class="song-info">
        <img class="track-img" width="90" height="90" [ngSrc]="item.track.album.images[1].url">
        <div class="album-artist">
          <h2 class="name">{{item.track.name}}</h2>
          <span class="album-name">
            {{item.track.album.name}}
          </span>
          <div class="track-artist">
            @for(artist of item.track.artists; track artist.id){
            <span class="artist">{{artist.name}}</span>
            }
          </div>
        </div>
      </div>
      <span class="time">{{item.track.duration_ms | duration}}
      </span>
    </li>
    }
  </ul>
  }
  @if(tracksListQuery().isLoading()){
  <ul class="track-list track-list--skeleton">
    @for(item of skeletonLoadingArray; track item; let i = $index){
    <li class="track">
      <div class="song-info">
        <ngx-skeleton-loader animation="progress-dark" count="1"
          [theme]="{ height: '90px', width: '90px', 'background-color': '#2c2f31' }"></ngx-skeleton-loader>
        <div class="album-artist">
          <ngx-skeleton-loader animation="progress-dark" count="1"
            [theme]="{ height: '32px', width: '160px', 'background-color': '#2c2f31' }"></ngx-skeleton-loader>
          <ngx-skeleton-loader class="artist-skeleton" animation="progress-dark" count="2"
            [theme]="{ height: '16px', width: '150px', 'background-color': '#2c2f31' }"></ngx-skeleton-loader>
        </div>
      </div>
      <span class="time">
        <ngx-skeleton-loader animation="progress-dark" count="1"
          [theme]="{ height: '16px', width: '30px', 'background-color': '#2c2f31' }"></ngx-skeleton-loader>
      </span>
    </li>
    }
  </ul>
  }
</div>