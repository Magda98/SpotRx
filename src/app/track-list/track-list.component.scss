.track-list-container {
  display: flex;
  flex-direction: column;
  max-height: 65vh;
}

.track-list {
  display: flex;
  justify-content: center;
  flex-direction: column;
  margin: auto;
  width: 100%;
  padding: 0;

  @media screen and (min-width: 980px) {
    width: 80%;
  }

  .song-info {
    display: flex;
  }

  .track {
    display: flex;
    justify-content: space-between;
    align-items: center;
    margin-bottom: 20px;
    padding: 8px 20px;
    border-radius: 10px;
    cursor: pointer;
    gap: 8px;
    transition: background-color 0.3s;

    &:hover {
      background-color: #2c2f31;
    }

    .track-img {
      width: 90px;
      height: 90px;
      object-fit: cover;
      margin-right: 10px;
      border-radius: 5px;
    }
  }

  &.track-list--skeleton {
    .song-info {
      gap: 16px;
    }

    .album-artist,
    .artist-skeleton {
      display: flex;
      flex-direction: column;
    }
  }
}

.paginator {
  margin-bottom: 16px;
}

.mat-paginator {
  margin: auto;
  width: 80%;
  background-color: #35393C;
  color: #fff;
}
