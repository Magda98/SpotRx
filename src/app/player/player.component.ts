import { PlayerService } from './../services/player.service';
import { Component, OnDestroy, OnInit } from '@angular/core';
import { MatSliderDragEvent } from '@angular/material/slider';
import {
  EMPTY,
  Subscription,
  distinctUntilChanged,
  finalize,
  map,
  switchMap,
  tap,
} from 'rxjs';
import { TrackService } from '../services/track.service';

@Component({
  selector: 'app-player',
  templateUrl: './player.component.html',
  styleUrls: ['./player.component.scss'],
})
export class PlayerComponent implements OnInit, OnDestroy {
  public playerState = this.playerService.getPlayerState();
  public player?: Spotify.PlaybackState;
  public isSavedTrack?: boolean;
  private subscription = new Subscription();
  private toggleTrackSubscription?: Subscription;
  private isFetching = false;

  constructor(
    public playerService: PlayerService,
    private trackService: TrackService
  ) {}

  ngOnInit(): void {
    this.subscription.add(
      this.playerState
        .pipe(
          tap((player) => {
            this.player = player;
          }),
          map((player) => player.track_window.current_track.id),
          distinctUntilChanged(),
          switchMap((id) => {
            if (id) return this.trackService.checkUserSavedTrack(id);
            return EMPTY;
          })
        )
        .subscribe((data) => {
          const [isSavedTrack] = data;
          this.isSavedTrack = isSavedTrack;
        })
    );
  }

  toggleSavedTrack() {
    if (!this.player?.track_window?.current_track?.id) return;

    if (this.isSavedTrack) {
      this.toggleTrackSubscription?.unsubscribe();
      return (this.toggleTrackSubscription = this.trackService
        .deleteTrack(this.player.track_window.current_track.id)
        .subscribe(() => {
          this.isSavedTrack = false;
        }));
    }

    this.toggleTrackSubscription?.unsubscribe();
    return (this.toggleTrackSubscription = this.trackService
      .saveTrack(this.player.track_window.current_track.id)
      .subscribe(() => {
        this.isSavedTrack = true;
      }));
  }

  setVolume(volumeSlider: MatSliderDragEvent) {
    this.playerService.setVolume(volumeSlider.value);
  }

  setPosition(positionSlider: MatSliderDragEvent) {
    this.playerService.setPosition(positionSlider.value);
  }

  toggleShuffle() {
    if (!this.player) return;
    if (this.isFetching) return;

    this.isFetching = true;
    this.player.shuffle = !this.player.shuffle;
    this.subscription.add(
      this.trackService
        .toggleShuffle(this.player.shuffle)
        .pipe(
          finalize(() => {
            this.isFetching = false;
          })
        )
        .subscribe()
    );
  }

  ngOnDestroy(): void {
    this.subscription.unsubscribe();
    this.toggleTrackSubscription?.unsubscribe();
  }
}
