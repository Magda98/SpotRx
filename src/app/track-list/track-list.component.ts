import { TrackService } from './../services/track.service';
import { PlayerService } from './../services/player.service';
import { Item } from './../interfaces/track';
import { Component, EventEmitter, Input, Output, Signal } from '@angular/core';
import { PageEvent } from '@angular/material/paginator';
import { toSignal } from '@angular/core/rxjs-interop';

@Component({
  selector: 'app-track-list',
  templateUrl: './track-list.component.html',
  styleUrls: ['./track-list.component.scss'],
})
export class TrackListComponent {
  @Input() tracksList!: Signal<Item[]>;
  @Input() total = 0;
  @Output() getNextPage = new EventEmitter<PageEvent>();
  currentPage = toSignal(this.trackService.currentPage);
  pageSize = this.trackService.pageSize;
  readonly skeletonLoadingArray = Array.from({ length: 6 }, () => null);

  constructor(
    private playerService: PlayerService,
    private trackService: TrackService
  ) {}

  play(index: number) {
    const queue = this.tracksList()?.map((item) => item.track.uri);
    if (queue) this.playerService.queue.next({ queue, index });
  }

  handlePageEvent(page: PageEvent) {
    this.trackService.currentPage.next(page.pageIndex);
    this.getNextPage.emit(page);
  }
}
