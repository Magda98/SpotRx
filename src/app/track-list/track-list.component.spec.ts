import { render, screen } from '@testing-library/angular';
import { TrackListComponent } from './track-list.component';
import { PlayerComponent } from '../player/player.component';
import {
  QueryClient,
  provideAngularQuery,
} from '@tanstack/angular-query-experimental';
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { MockBackendInterceptor } from '../../tests/mock-backend.interceptor';
import { savedTracks } from '../../tests/mocks';
import { PlayerService } from '../services/player.service';
import { HttpClientTestingModule } from '@angular/common/http/testing';

describe('TrackListComponent', () => {
  test('should render', async () => {
    const component = await render(TrackListComponent, {
      providers: [PlayerService],
      imports: [HttpClientTestingModule],
      componentInputs: {
        tracksList: savedTracks,
        isLoading: false,
        total: savedTracks.total,
      },
    });

    await screen.findByAltText('Saturn');
    expect(component.container).toMatchSnapshot();
  });
});
