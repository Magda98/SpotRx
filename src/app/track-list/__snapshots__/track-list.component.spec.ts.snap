// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrackListComponent should render 1`] = `
<div
  id="root0"
>
  <div
    class="track-list-container"
  >
    <mat-paginator
      aria-label="Select page"
      class="mat-mdc-paginator paginator"
      role="group"
    >
      <div
        class="mat-mdc-paginator-outer-container"
      >
        <div
          class="mat-mdc-paginator-container"
        >
          <div
            class="mat-mdc-paginator-page-size"
          >
            <div
              class="mat-mdc-paginator-page-size-label"
            >
               Items per page: 
            </div>
            
            <div
              class="mat-mdc-paginator-page-size-value"
            >
              6
            </div>
            
          </div>
          
          <div
            class="mat-mdc-paginator-range-actions"
          >
            <div
              aria-live="polite"
              class="mat-mdc-paginator-range-label"
            >
               1 – 6 of 373 
            </div>
            
            <button
              aria-label="Previous page"
              class="mat-mdc-tooltip-trigger mat-mdc-paginator-navigation-previous mdc-icon-button mat-mdc-icon-button mat-mdc-button-disabled _mat-animation-noopable mat-unthemed mat-mdc-button-base mat-mdc-tooltip-disabled"
              disabled="true"
              mat-icon-button=""
              mat-ripple-loader-centered=""
              mat-ripple-loader-class-name="mat-mdc-button-ripple"
              mat-ripple-loader-disabled=""
              mat-ripple-loader-uninitialized=""
              type="button"
            >
              <span
                class="mat-mdc-button-persistent-ripple mdc-icon-button__ripple"
              />
              <svg
                aria-hidden="true"
                class="mat-mdc-paginator-icon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                />
              </svg>
              <span
                class="mat-mdc-focus-indicator"
              />
              <span
                class="mat-mdc-button-touch-target"
              />
            </button>
            
            <button
              aria-label="Next page"
              class="mat-mdc-tooltip-trigger mat-mdc-paginator-navigation-next mdc-icon-button mat-mdc-icon-button _mat-animation-noopable mat-unthemed mat-mdc-button-base"
              mat-icon-button=""
              mat-ripple-loader-centered=""
              mat-ripple-loader-class-name="mat-mdc-button-ripple"
              mat-ripple-loader-uninitialized=""
              type="button"
            >
              <span
                class="mat-mdc-button-persistent-ripple mdc-icon-button__ripple"
              />
              <svg
                aria-hidden="true"
                class="mat-mdc-paginator-icon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                />
              </svg>
              <span
                class="mat-mdc-focus-indicator"
              />
              <span
                class="mat-mdc-button-touch-target"
              />
            </button>
            
            
          </div>
        </div>
      </div>
    </mat-paginator>
    <ul
      class="track-list"
    >
      <li
        class="track"
      >
        <div
          class="song-info"
        >
          <img
            alt="Saturn"
            class="track-img"
            fetchpriority="auto"
            height="90"
            loading="lazy"
            ng-img="true"
            src="https://i.scdn.co/image/ab67616d00001e02d70916ee9e40c90380ba5f07"
            width="90"
          />
          <div
            class="album-artist"
          >
            <h2
              class="name"
            >
              Saturn
            </h2>
            <span
              class="album-name"
            >
               Saturn 
            </span>
            <div
              class="track-artist"
            >
              <span
                class="artist"
              >
                SZA
              </span>
              
            </div>
          </div>
        </div>
        <span
          class="time"
        >
          3:06 
        </span>
      </li>
      <li
        class="track"
      >
        <div
          class="song-info"
        >
          <img
            alt="River"
            class="track-img"
            fetchpriority="auto"
            height="90"
            loading="lazy"
            ng-img="true"
            src="https://i.scdn.co/image/ab67616d00001e028cae5034066af45cdfbc4266"
            width="90"
          />
          <div
            class="album-artist"
          >
            <h2
              class="name"
            >
              River
            </h2>
            <span
              class="album-name"
            >
               Church Of Scars 
            </span>
            <div
              class="track-artist"
            >
              <span
                class="artist"
              >
                Bishop Briggs
              </span>
              
            </div>
          </div>
        </div>
        <span
          class="time"
        >
          3:36 
        </span>
      </li>
      <li
        class="track"
      >
        <div
          class="song-info"
        >
          <img
            alt="Daggers"
            class="track-img"
            fetchpriority="auto"
            height="90"
            loading="lazy"
            ng-img="true"
            src="https://i.scdn.co/image/ab67616d00001e0210311a15fdac768608fe224d"
            width="90"
          />
          <div
            class="album-artist"
          >
            <h2
              class="name"
            >
              Daggers
            </h2>
            <span
              class="album-name"
            >
               Daggers 
            </span>
            <div
              class="track-artist"
            >
              <span
                class="artist"
              >
                Croosh
              </span>
              
            </div>
          </div>
        </div>
        <span
          class="time"
        >
          2:19 
        </span>
      </li>
      <li
        class="track"
      >
        <div
          class="song-info"
        >
          <img
            alt="Bécane - A COLORS SHOW"
            class="track-img"
            fetchpriority="auto"
            height="90"
            loading="lazy"
            ng-img="true"
            src="https://i.scdn.co/image/ab67616d00001e027dfba3f3ae607f250cdd969c"
            width="90"
          />
          <div
            class="album-artist"
          >
            <h2
              class="name"
            >
              Bécane - A COLORS SHOW
            </h2>
            <span
              class="album-name"
            >
               Bécane - A COLORS SHOW 
            </span>
            <div
              class="track-artist"
            >
              <span
                class="artist"
              >
                Yamê
              </span>
              
            </div>
          </div>
        </div>
        <span
          class="time"
        >
          3:02 
        </span>
      </li>
      <li
        class="track"
      >
        <div
          class="song-info"
        >
          <img
            alt="Supergirl - Radio Edit"
            class="track-img"
            fetchpriority="auto"
            height="90"
            loading="lazy"
            ng-img="true"
            src="https://i.scdn.co/image/ab67616d00001e0260e18fbad67fc3c48608d03a"
            width="90"
          />
          <div
            class="album-artist"
          >
            <h2
              class="name"
            >
              Supergirl - Radio Edit
            </h2>
            <span
              class="album-name"
            >
               Supergirl (Radio Edit) 
            </span>
            <div
              class="track-artist"
            >
              <span
                class="artist"
              >
                Anna Naklab
              </span>
              <span
                class="artist"
              >
                Alle Farben
              </span>
              <span
                class="artist"
              >
                YouNotUs
              </span>
              
            </div>
          </div>
        </div>
        <span
          class="time"
        >
          3:32 
        </span>
      </li>
      <li
        class="track"
      >
        <div
          class="song-info"
        >
          <img
            alt="Santa, Can’t You Hear Me"
            class="track-img"
            fetchpriority="auto"
            height="90"
            loading="lazy"
            ng-img="true"
            src="https://i.scdn.co/image/ab67616d00001e021fa667a7a53c7072fed91a17"
            width="90"
          />
          <div
            class="album-artist"
          >
            <h2
              class="name"
            >
              Santa, Can’t You Hear Me
            </h2>
            <span
              class="album-name"
            >
               When Christmas Comes Around... 
            </span>
            <div
              class="track-artist"
            >
              <span
                class="artist"
              >
                Kelly Clarkson
              </span>
              <span
                class="artist"
              >
                Ariana Grande
              </span>
              
            </div>
          </div>
        </div>
        <span
          class="time"
        >
          4:02 
        </span>
      </li>
      
    </ul>
    
    
  </div>
</div>
`;
